From 5db8b282787849982fc42d1d1bf67dbc3640cbcc Mon Sep 17 00:00:00 2001
From: Richard Gong <richard.gong@intel.com>
Date: Fri, 19 Jun 2020 12:50:46 -0500
Subject: [PATCH 094/151] crypto: intel_fcs: update to properly handle NULL
 pointer

commit  023c37a7359be6ea83bc60c977db15a8dfe9f9e9 from
https://github.com/altera-opensource/linux-socfpga.git
branch is socfpga-5.4.64-lts

When Arm Trust Firmware (ATF) rejects Linux's request on Secure Data
Object Storage (SDOS), ATF doesn't return any error code from mailbox.
Kernel crypto service driver uses NULL pointer dereference which cause
kernel panic.

Update kernel crypto service driver to properly handle NULL pointer
dereference to prevent kernel panic.

Signed-off-by: Richard Gong <richard.gong@intel.com>
Signed-off-by: Meng Li <Meng.Li@windriver.com>
Integrated-by: Jun Zhang <jun.zhang@windriver.com>
---
 drivers/crypto/intel_fcs.c | 10 +++++++---
 1 file changed, 7 insertions(+), 3 deletions(-)

diff --git a/drivers/crypto/intel_fcs.c b/drivers/crypto/intel_fcs.c
index 15bc62b62f56..56574d74ce93 100644
--- a/drivers/crypto/intel_fcs.c
+++ b/drivers/crypto/intel_fcs.c
@@ -48,15 +48,19 @@ static void fcs_data_callback(struct stratix10_svc_client *client,
 			     struct stratix10_svc_cb_data *data)
 {
 	struct intel_fcs_priv *priv = client->priv;
-	unsigned int *status = (unsigned int *)data->kaddr1;
 
 	if (data->status == BIT(SVC_STATUS_OK)) {
 		priv->status = 0;
 		priv->kbuf = data->kaddr2;
 		priv->size = *((unsigned int *)data->kaddr3);
+	} else if (data->status == BIT(SVC_STATUS_ERROR)) {
+		priv->status = *((unsigned int *)data->kaddr1);
+		dev_err(client->dev, "error, mbox_error=0x%x\n", priv->status);
+		priv->kbuf = NULL;
+		priv->size = 0;
 	} else {
-		dev_err(client->dev, "failed, mbox_error=0x%x\n", *status);
-		priv->status = *status;
+		dev_err(client->dev, "rejected\n");
+		priv->status = -EINVAL;
 		priv->kbuf = NULL;
 		priv->size = 0;
 	}
-- 
2.26.1

